src/hotspot/share/prims/jvmtiEnv.cpp

```diff
@@ -1190,9 +1190,15 @@ JvmtiEnv::StopThread(jthread thread, jobject exception) {
 
   jvmtiError err = get_threadOop_and_JavaThread(tlh.list(), thread, &java_thread, &thread_oop);
 
-  if (thread_oop != nullptr && thread_oop->is_a(vmClasses::BaseVirtualThread_klass())) {
-    // No support for virtual threads (yet).
-    return JVMTI_ERROR_UNSUPPORTED_OPERATION;
+  bool is_virtual = thread_oop != nullptr && thread_oop->is_a(vmClasses::BaseVirtualThread_klass());
+
+  if (is_virtual && !is_JavaThread_current(java_thread, thread_oop)) {
+    if (!JvmtiVTSuspender::is_vthread_suspended(thread_oop)) {
+      return JVMTI_ERROR_THREAD_NOT_SUSPENDED;
+    }
+    if (java_thread == nullptr) { // unmounted virtual thread
+      return JVMTI_ERROR_OPAQUE_FRAME;
```

Comment:
Where is the check for "suspended at an event" that otherwise results in `JVMTI_ERROR_OPAQUE_FRAME`?

Reply:
The JVMTI `StopThread` spec has this description:
> The StopThread function may be used to send an asynchronous
> exception to a virtual thread when it is suspended at an event.
> An implementation may support sending an asynchronous exception
> to a suspended virtual thread in other cases.
> . . .
> JVMTI_ERROR_OPAQUE_FRAME:
> The thread is a suspended virtual thread and the implementation 
> was unable to throw an asynchronous exception from this frame.

This update supports all suspended mounted cases of virtual threads and returns OPAQUE_FRAME only if target virtual thread is suspended when unmounted.

Reply:
What does "suspended at an event" mean? As a programmer trying to use this how am I supposed to know when it can be used without getting an error?

I find it very surprising that the error would occur with an unmounted thread - having a VT throw when it was remounted seems the most natural way to implement this.

Reply:
I think "suspended at an event" is okay. It means the callback for an event has been triggered and the agent suspended the thread. The typical use-case for JVMTI StopThread is when at a breakpoint or when single stepping and the user asks the debugger to throw some exception so that the code's handling of the exception has been debugged/tested. Debugger and JDWP agent aside, I don't know if there are other agents using this JVMTI function. If there are other and they call this function on some random virtual thread at some random time then the function will fail.

One other point around this is that the plan is to have StopThread, ForceEarlyReturn, PopFrame and SetLocalXXX work more consistently. Right now, SetLocalXXX minimally requires a virtual thread suspended at a breakpoint or single step event. The minimum support can be broader to be suspended at any event.

Reply:
Alan, thank you for explaining this.
Resolving this conversation.
David, please, reopen it if it is no okay with you.